### ISSUE TITLE
Hi, Ludwig 0.3  and Tensoflow 2.3.1 versions are not compatible, any recent updates on the same?

### ISSUE THREAD

**Issue by jithincheriyan** [uber/ludwig#997]
**Describe the bug**

![image](https://user-images.githubusercontent.com/52187221/98302587-860de800-2021-11eb-9307-8d0c45f2ec70.png)

**Environment 
 - OS: Windows 10
- Python version 3.7
- Ludwig version 0.3

**Comment by w4nderlust** [comment]
Hi @jithincheriyan , trying to understand a few things:
- what makes you think that the error you are getting has to do with TF 2.3.1 specifically? We haven't encountered any problem with it so far.
- Can you provide a reproducible example so that we can look into it? A screenshot without an explanation of how it was obtained doesn't allow us to fix a problem or to spot the problem in what you are doing.

**Comment by jithincheriyan** [comment]
H i @w4nderlust ,

I tried Ludwig versions 0.2.2.8 and 0.3 with Tensorflow versions 2.3.1. But in both cases, it ended up in some errors. Finally I had to comedown to Ludwig 0.2.2.8 with TF 1.15.4 to make it work. 

I have uploaded the example that I have been working on for your reference.


[Ludwig Classifier.txt](https://github.com/uber/ludwig/files/5498429/Ludwig.Classifier.txt)

**Comment by w4nderlust** [comment]
By looking at your file I have the suspicion that you trained a model previously and now you want to use it with a different version of Ludwig. In particular with v0.3 that does not work, because of the different TF version and because of many other changes.

So I suggest to retrain the model, and I took the libery to adapt your code to v0.3, let me know if it works.

```
import os
import yaml
from ludwig.api import LudwigModel
import gzip,io
import csv
import pandas
import codecs
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'


model_definition = {'input_features': [{'name': 'Comment_Text', 'type': 'text'}],
                   'output_features': [{'name': 'Personal', 'type': 'category'},
                                       {'name': 'Racial', 'type': 'category'},
                                       {'name': 'Swearing', 'type': 'category'},
                                       {'name': 'Normal', 'type': 'category'}]}

# Training
path="H:\\ML_Fine_Grained_V3.csv"
ludwig_model = LudwigModel(model_definition)
stats = ludwig_model.train(dataset=path)  #needed
print(stats)   #needed
predictions = ludwig_model.predict(dataset="H:\\ML_Training_Data_To_Test.csv")    # needed if predictions needed


# Testing
ludwig_model = LudwigModel.load("H:\\results\\Fine_Grained_Ludwig_Model_1\\model")     # Already trained model
predictions, test_stats = ludwig_model.test(dataset="H:\\Test.csv")

#predictions.to_csv('H:\\Test_Predictions.tsv',sep='\t', encoding='utf-8') #Needed if large predictions have to be done
print(predictions)

# print(ludwig.api.kfold_cross_validate(10, model_definition=model_definition, dataset="H:\\temp.csv"))
```

**Comment by jithincheriyan** [comment]
Hi @w4nderlust ,

Yes, you are right, I trained the model using 0.2.2.8 and the present model is 0.3. 
Still I tried a new training with the code you had provided with TF 2.3.1 and got the following error. 

![image](https://user-images.githubusercontent.com/52187221/98324279-b91ba000-2050-11eb-8ecf-806444a0251e.png)

**Comment by w4nderlust** [comment]
The line that is giving you an error is not present in my code. I replaced all the dataset_csv with dataset. Make sure to actually use my code or to fix your code according to the changes in v0.3 ;)

**Comment by jithincheriyan** [comment]
Hi,

Sorry for the mistake, ran the previous file itself. Now your code started training with Ludwig 0.3 and TF 2.3.1

Thank you for the support. I will update you in case of any error on training or testing.

**Comment by jithincheriyan** [comment]
Hi,

The training went well and the model got stored in the "results" directory. However, while testing, came across the following error. 

![image](https://user-images.githubusercontent.com/52187221/98332330-58499300-2063-11eb-9aeb-e902a1c753aa.png)

Any suggestions....?

**Comment by w4nderlust** [comment]
yes, sorry, `test` was replaced with `evaluate`.

**Comment by jithincheriyan** [comment]
Thank you, it worked, though the predicted results are not good as the previous version.

Thank you for the support

**Comment by w4nderlust** [comment]
> Thank you, it worked, though the predicted results are not good as the previous version.

That's unexpected. it should work as well if not better because many improvement we did. I suggest you to adjust some parameters (the learning rate in particular) as it may influence training substantially, and we fixed a few things about it.

**Comment by jithincheriyan** [comment]
Hi,

I would like to know whether visualization is there for LudwigAPI. I tried the same for learning rate optimization but cannot find a function for that. If yes, will you please be able to inform me along some examples.

**Comment by w4nderlust** [comment]
Sorry I'm not sure what do you mean by learning rate optimization. I guess you mean hyperopt, in which case here's the user guide section for you: https://ludwig-ai.github.io/ludwig-docs/user_guide/#hyper-parameter-optimization
Regarding the visualization, please check the user guide section: https://ludwig-ai.github.io/ludwig-docs/user_guide/#visualizations and the API section: https://ludwig-ai.github.io/ludwig-docs/api/visualization/

**Comment by jithincheriyan** [comment]
Hi, 
Thank you for the directions.
I just wanted to plot the loss function to adjust the learning rate for better results. I believe the present API documentation is for Ludwig 0.3, but while coding, it  does not come up with visualization function. 

Will you please be able to mention how visualization can be done for loss function for the model detailed below.

from ludwig.api import LudwigModel
model_definition = {'input_features': [{'name': 'Comment_Text', 'type': 'text'}],
                   'output_features': [{'name': 'Personal', 'type': 'category'},
                                       {'name': 'Racial', 'type': 'category'},
                                       {'name': 'Swearing', 'type': 'category'},
                                       {'name': 'Normal', 'type': 'category'}]}
#Training
path="H:\\ML_Fine_Grained_V3_Preprocessed.csv"
ludwig_model = LudwigModel(model_definition)
stats = ludwig_model.train(dataset=path) 
print(stats)

**Comment by w4nderlust** [comment]
I'm not sure what you mean by "while coding, it does not come up with visualization function".
For visualizing the loss over time you cna use the `learning_curves` function: https://ludwig-ai.github.io/ludwig-docs/api/visualization/#learning_curves or open a tensor board.

**Comment by jithincheriyan** [comment]
Hi @w4nderlust ,

I resolved the issues by updating and reinstalling Ludwig (0.3) and Tensorflow (2.3.1). 

But now I am facing another issue of " Intra op parallelism cannot be modified after initialization".
Besides Ludwig, I also use pipeline from transformers for another related task.
These two, Ludwig and transformers works fine independently, but shows the following error when combined together as a program.
Will you please find the attached error message and make some suggestions?

![image](https://user-images.githubusercontent.com/52187221/99202573-802fb800-2814-11eb-8dfa-db20efb252ff.png)

**Comment by w4nderlust** [comment]
@jithincheriyan glad to hear you solved the issue. The screen that you posted is not sufficient for me to solve the issue, I need more information. One guess is that you are either playing around with tensorflow config before using Ludwig in your script or you are trying to change those reproducibility parameters within ludwig in a way that is not supported, but without knowing the code i can't really tell unfortunately.

**Comment by jithincheriyan** [comment]
Hi,

Thank you for the response. 

I used one classifier using Ludwig, the code of the module is:

def Ludwig_classifier(Comment):
    Comment = {"Comment_Text": Comment, "Personal": '', "Racial": '', "Swearing": '', "Normal": ''}
    Df = pd.DataFrame(columns=["Comment_Text", "Personal", "Racial", "Swearing", "Normal"])
    Df = Df.append(Comment, ignore_index=True)

    model_definition={'input_features': [{'name': 'Comment_Text', 'type': 'text'}],
                      'output_features': [{'name': 'Personal', 'type': 'category'},
                                          {'name': 'Racial', 'type': 'category'},
                                          {'name': 'Swearing', 'type': 'category'},
                                          {'name': 'Normal', 'type': 'category'}]}

    ludwig_model = LudwigModel.load("H:\\results\\Fine_Grained_Ludwig_Model_2\\model") # Already trained model
    predictions,_=ludwig_model.predict(dataset=Df,model_definition=model_definition)

    return (predictions['Personal_predictions'][0],
            predictions['Racial_predictions'][0],
            predictions['Swearing_predictions'][0],
            predictions['Normal_predictions'][0])

Also, I use another module using transformer pipeline. Please find the following module

from transformers import pipeline
unmasker = pipeline('fill-mask', model='bert-base-uncased')
def Comment_Paraphraser(Comment):
    p = re.compile(r'[\w!]+')
    for location in Find_Regex(Comment):
        Sub_String=Comment[location['Start']:location['End']]
        for i in p.finditer(Comment):
            if Sub_String in i.group():
                Comment=Comment[:i.start()]+'[MASK]'+Comment[i.end():]
                results=unmasker(Comment)
                Comment=results[0]['sequence'][6:-6]    # to trim extra characters added as part of masking
                break
    return (Comment)


These two modules work fine when ran as separate programs, but not when combined together as separate modules of a program, which is my concern.

**Comment by w4nderlust** [comment]
I can't be sure, but I guess they both try to set come tensorflow parameters that can be set only independently, so i suggest you to use only one. Anyway, all the important encoders in transformers are already available in Ludwig. If you specify `{'name': 'Comment_Text', 'type': 'text', encoder: bert}` you will be using bert, and can also specify the specific `bert-base-uncased` if you want, look here for more details: https://ludwig-ai.github.io/ludwig-docs/user_guide/#text-input-features-and-encoders

**Comment by jithincheriyan** [comment]
Hi,
Thank you for the suggestion. 
I changed both the encoders to _bert-base-uncased_ and it works.
Thank you so much for the help.
